### ISSUE TITLE
Small error/logging refactor

### ISSUE THREAD

**Issue by Ericson2314** [NixOS/nix#3841]
Currently There is an `ErrorInfo` struct which

1. Contains a verbosity level
2. Is a field of `BaseError`

I think this causes problems:

1. It doesn't make sense to associate exceptions with verbosity levels: an exception is a failure that aborts the computation. (It's the *handler* / catch block that decide whether something can keep going and isn't fatal.)

2. There is no way to render an error different based on the verbosity level: @edolstra and I on a call talked bout making HTTP error response bodies show up with higher log levels, this doesn't appear very easy.

Here is what I tentatively  propose:

1. `ErrorInfo` is really about logging messages, which may or may not have anything to do with exceptions/control flow. It should be renamed to `LogEntry`, and put in `logging,hh`.
    - `error.hh` would import `logging.hh` instead of the other way around
    - `enum Verbosity` can also move back

2. Rather than storing an `ErrorInfo`, `BaseError` should have virtual method:
    ```c++
    // Render the exception into a log entry containing more or less information bes
    virtual LogEntry render(Verbosity v) = 0;
    ```
    which will allow exceptions to be rendered with different amount of information based on the `-vvv`.

Part 1 I definitely like, part 2 I am not sure about because the `render` should always return a `lvlError`, right? so the `Verbosity` parameter (how much information is in the error) is used rather incongruously.

**Comment by bburdette** [comment]
I don't object to renaming ErrorInfo, or moving it around.  It IS about the logging.  To me the main advantage with ErrorInfo is we're getting structured data to the logger rather than just strings.  

Right now, though, the ErrorInfo is not extensible.  Extensibility is just an extra complication if its not really needed, but maybe its necessary.  The ErrorInfo has structured data in it for Nix language errors, but it doesn't have special fields for anything else - http in this instance.  

Having a render function in exceptions would mean that you would need to store all the information that goes into an ErrorInfo (or 'LogEntry') in the exception, so that you could build an ErrorInfo later and pass it to the logger.  Just storing the ErrorInfo in the first place seems like a clean way to go to me, assuming it can contain the data you care about.

If you only care about rendering http response codes into a hint string in the log, you can do it by checking verbosity when creating the ErrorInfo.  If verbosity is high, put the code into the hint, if its low leave it out.  That doesn't require code changes in the ErrorInfo/Logger.

However, if you wanted to add an http response code field into a json log entry in the json logger, there doesn't seem to be a good way to do that now.  You'd be stuck having to parse the code out of the text string in the hint.  Some ideas for adding extra data to the json output:

1) Use inheritance.  Have a render-to-json method in the ErrorInfo.  Override this in a descendant to call the parent and add more json fields, for http or whatever else comes along.

2) Add a name-value map in the ErrorInfo.  The json logger would iterate through these and add them as fields to the json.  

3) Add an 'extra data' string field into ErrorInfo that's json.  When we make json, we add the extra data in.

I haven't fully thought these through, just throwing them out for discussion.  I like the simplicity of having just one ErrorInfo and not a family of descendants.  On the other hand, 2 and 3 mean you'd be adding data fields that aren't used if json logging is disabled.

**Comment by Ericson2314** [comment]
Yeah so for me `ErrorInfo` is still the semi-structured *result* of rendering. The underlying data is stored in the exception which no longer has an `ErrorInfo` field. A more structured way to present that information either is left as future work---I agree sub-classing `ErrorInfo` doesn't feel right.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
